cdap-data-fabric/src/main/java/co/cask/cdap/data2/registry/UsageTable.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright © 2015-2019 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.data2.registry;
+
+import co.cask.cdap.api.dataset.lib.CloseableIterator;
+import co.cask.cdap.proto.id.ApplicationId;
+import co.cask.cdap.proto.id.DatasetId;
+import co.cask.cdap.proto.id.ProgramId;
+import co.cask.cdap.spi.data.StructuredRow;
+import co.cask.cdap.spi.data.StructuredTable;
+import co.cask.cdap.spi.data.StructuredTableContext;
+import co.cask.cdap.spi.data.table.field.Field;
+import co.cask.cdap.spi.data.table.field.Fields;
+import co.cask.cdap.spi.data.table.field.Range;
+import co.cask.cdap.store.StoreDefinition;
+import com.google.common.annotations.VisibleForTesting;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Store program -> dataset/stream usage information.
+ */
+public class UsageTable {
+
+  private final StructuredTable table;
+
+  public UsageTable(StructuredTableContext context) {
+    this.table = context.getTable(StoreDefinition.UsageStore.USAGES);
+  }
+
+  @VisibleForTesting
+  void deleteAll() throws IOException {
+    table.deleteAll(Range.all());
+  }
+
+  /**
+   * Registers usage of a dataset by a program.
+   * @param programId program
+   * @param datasetInstanceId dataset
+   */
+  public void register(ProgramId programId, DatasetId datasetInstanceId) throws IOException {
+    List<Field<?>> fields = getProgramPrefix(programId);
+    fields.add(Fields.stringField(StoreDefinition.UsageStore.DATASET_FIELD, datasetInstanceId.getDataset()));
+    // TODO: CDAP-14947 For whatever reason, an index cannot be a primary key...
```

Comment:
Can you add more details to the JIRA? Is it happening to both nosql and sql implementation, is scan giving wrong results or throw some exception?